test/forms/inputs/cloudinary-file-input/cloudinary-uploader.test.js

```diff
@@ -115,232 +139,373 @@ describe('cloudinaryUploader', () => {
     // eslint-disable-next-line no-undef
     process.env.CLOUDINARY_BUCKET = 'bar'
     const Wrapped = () => <h1>Hi</h1>
-    const Wrapper = cloudinaryUploader({ apiAdapter: props.apiAdapter })(Wrapped)
+    const Wrapper = cloudinaryUploader({ apiAdapter: props.apiAdapter })(
+      Wrapped
+    )
     expect(() => render(<Wrapper />)).not.toThrow()
   })
 
   test('can receive options as props', () => {
     const Wrapped = () => <h1>Hi</h1>
     const Wrapper = cloudinaryUploader()(Wrapped)
-    expect(() => render(
-      <Wrapper cloudName="foo" bucket="bar" apiAdapter={() => {}} />
-    )).not.toThrow()
+    expect(() =>
+      render(<Wrapper cloudName="foo" bucket="bar" apiAdapter={() => {}} />)
+    ).not.toThrow()
   })
 
   test('adds upload props to component', () => {
     const Wrapped = jest.fn(() => <h1>Hi</h1>)
     const Wrapper = cloudinaryUploader(props)(Wrapped)
     render(<Wrapper />)
-    expect(Wrapped).toHaveBeenCalledWith(expect.objectContaining({
-      upload: expect.any(Function),
-      uploadStatus: expect.any(String),
-    }), {})
+    expect(Wrapped).toHaveBeenCalledWith(
+      expect.objectContaining({
+        upload: expect.any(Function),
+        uploadStatus: expect.any(String),
+      }),
+      {}
+    )
   })
 
-  test('sends the api request with the correct options', () => {
-    const Wrapped = jest.fn(() => <h1>Hi</h1>)
+  test('sends the api request with the correct options', async () => {
+    let response
+    const Wrapped = (props) => (
+      <AsyncWrapper
+        onSuccess={(res) => {
+          response = res
+        }}
+        fileData={fileData}
+        file={file}
+        {...props}
+      />
+    )
     const Wrapper = cloudinaryUploader(props)(Wrapped)
+    const user = userEvent.setup()
     render(<Wrapper />)
 
-    const { upload } = Wrapped.mock.calls[0][0]
-
-    return upload(fileData, file).then((response) => {
-      const { uploadStatus } = Wrapped.mock.calls[2][0]
```

Comment:
This approach didn't work when we switched to React 18. I think something happened related to React's introduction of automated batching and `act`. Instead of going from `''` to `'uploading'` to `'upload-success'`, there would only be one state update directly to `upload-success`.

The way that I've rewritten the tests is much closer to how the user experiences the functionality, which aligns well with the RTL philosophy.

Reply:
BTW I tested these changes on React 17 before bumping to make sure I wasn't introducing regressions

Reply:
Always happy to remove `mock.calls[][][]` 😄

But yes, it seems to be related to automatic batching, from my understanding of [the](https://react.dev/blog/2022/03/29/react-v18#new-feature-automatic-batching) [docs](https://github.com/reactwg/react-18/discussions/21).